core/mount/mount_linux_test.go

```diff
@@ -198,6 +206,120 @@ func TestUnmountRecursive(t *testing.T) {
 	}
 }
 
+func TestDoPrepareIDMappedOverlay(t *testing.T) {
+	testutil.RequiresRoot(t)
+
+	k512 := kernel.KernelVersion{Kernel: 5, Major: 12}
+	ok, err := kernel.GreaterEqualThan(k512)
+	require.NoError(t, err)
+	if !ok {
+		t.Skip("GetUsernsFD requires kernel >= 5.12")
+	}
+
+	usernsFD, err := getUsernsFD(testUIDMaps, testGIDMaps)
+	require.NoError(t, err)
+	defer usernsFD.Close()
+
+	type testCase struct {
+		name              string
+		injectUmountFault bool
+	}
+
+	tcases := []testCase{
+		{
+			name:              "normal",
+			injectUmountFault: false,
+		},
+		{
+			name:              "umount-fault",
+			injectUmountFault: true,
+		},
+	}
+
+	for _, tc := range tcases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Reserve an OS thread with full capabilities for cleanup.
+			testCompleteChan := make(chan struct{})
+			cleanup := func() {}
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
+
+			// Force the go runtime to run the actual test on a separate thread, where we may drop caps.
+			go func() {
+				defer close(testCompleteChan)
+				fakeLowerDirsDir := t.TempDir()
+				lowerDirs := []string{filepath.Join(fakeLowerDirsDir, "lower1"), filepath.Join(fakeLowerDirsDir, "lower2")}
+				for _, dir := range lowerDirs {
+					require.NoError(t, os.Mkdir(dir, 0755))
+					require.NoError(t, os.WriteFile(filepath.Join(dir, filepath.Base(dir)), []byte("foo"), 0644))
+				}
+
+				remountsLocation := t.TempDir()
+
+				tmpLowerDirs, c, err := doPrepareIDMappedOverlay(remountsLocation, lowerDirs, int(usernsFD.Fd()))
+				cleanup = c
+				require.NoError(t, err)
+				require.Len(t, tmpLowerDirs, len(lowerDirs))
+
+				lowerContents := make([][]byte, len(lowerDirs))
+
+				for i, dir := range lowerDirs {
+					correspondingRemount := tmpLowerDirs[i]
+					filename := filepath.Base(dir)
+
+					expectedFile, err := os.ReadFile(filepath.Join(dir, filename))
+					require.NoError(t, err, "reading comparison test fixture file")
+					lowerContents[i] = expectedFile
+
+					actualFile, err := os.ReadFile(filepath.Join(correspondingRemount, filename))
+					require.NoError(t, err, "reading file in temporary remount")
+
+					assert.Equal(t, expectedFile, actualFile, "file content in temporary remount")
+				}
+
+				if tc.injectUmountFault {
+					runtime.LockOSThread()
+					// Never unlock in this goroutine, ensures the thread isn't reused.
+					// Thus we don't have to worry about restoring the caps.
+					ourCaps, err := capability.NewPid2(0)
```

Comment:
I think this is quite complicated. This is basically to test the cleanup function, right?

I have a WIP change to just move the cleanup to a different function, so we can easily unit-test it. It's still a WIP, if you want, I can share it with you.

But wouldn't have the cleanup be a different function simplify the complexity of the test here? Or am I missing something?

Reply:
> I think this is quite complicated.

I think so too.

> This is basically to test the cleanup function, right?

Yes, specifically to generate a fault with unmount even when it has MNT_DETACH and verify that we don't remove the underlying files. We could also just not test that, it's vanishingly unlikely with MNT_DETACH present and we have multiple safeguards at this point to prevent it.

I'm also open to doing it in a more unit test / dependency injected way, I was just trying to keep the outside _test.go change size small. Happy to take a look at your approach!